src/containers/CustomFunctionModal.tsx

```diff
@@ -0,0 +1,212 @@
+    try {
+      const context: any = {}
+      functionVariables.forEach(variable => {
+        context[variable] = 2
+      })
+      parser(func, context)
+    } catch (err) {
+      console.error(err)
+      setGenericError()
+      return false
+    }
+
+    return true
+  }
+
+  onSave = () => {
+    const { customFunction, deactivateModal, createFunction, setFunction } = this.props
+    const { validateName, validateFunc } = this
+    const { name, func } = this.state
+
+    if (!validateName() || !validateFunc()) {
+      return
+    }
+
+    if (!customFunction) {
+      createFunction(name, func)
+    } else {
+      setFunction(customFunction.id, name, func)
+    }
+
+    deactivateModal()
+  }
+
+  onDelete = () => {
+    const { customFunction, deactivateModal, deleteFunction } = this.props
+
+    deleteFunction(customFunction.id)
+    deactivateModal()
+  }
+
+  render() {
+    const { name, func, nameError, funcError } = this.state
+    const { customFunction, deactivateModal } = this.props
+    const { onSave, onDelete } = this
+
+    const footer = (
+      <div className="buttons">
+        {customFunction && customFunction.id ? (
+          <button type="button" className="button is-danger" onClick={onDelete}>{t`Delete`}</button>
+        ) : (
+          <div />
+        )}
+        <span className="button-group">
+          <button type="button" className="button is-dark" onClick={deactivateModal}>{t`Cancel`}</button>
+          <button type="button" className="button is-primary" onClick={onSave}>{t`Save`}</button>
+        </span>
+      </div>
+    )
+
+    return (
+      <ModalCard
+        title={t`Add Function`}
+        footer={footer}
+        className="custom-function-modal"
+        onHide={deactivateModal}
+        active
+      >
+        <label>
+          {t`Name`}
+          <div className="error">{nameError}</div>
+          <input
+            value={name}
+            ref={this.nameRef}
+            onChange={e => {
+              this.setState({ name: e.target.value })
+              this.setState({ nameError: '' })
+            }}
+            onKeyPress={e => {
+              if (e.key === 'Enter' && this.funcRef.current) {
+                this.funcRef.current.focus()
+              }
+            }}
+          />
+        </label>
+        <label>
+          {t`Function`}
+          <div className="error">{funcError}</div>
+          <textarea
+            value={func}
+            ref={this.funcRef}
+            onChange={e => {
+              // strips everything not understood by the parser except spaces
+              const formatted = e.target.value.replace(/[^A-Za-z0-9 */_+\-(.)]/g, '')
+              this.setState({ func: formatted })
+            }}
+            onKeyPress={e => {
+              if (e.key === 'Enter') {
+                return onSave()
+              }
+              this.setState({ funcError: '' })
+            }}
```

Comment:
Since this is a textarea, we probably don't want to do stuff when the user presses "Enter", which they'll do if they're spreading a function across several lines.